### PR TITLE
*: support writing rows with checksum values

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #42747

Problem Summary:

As decribed in #42747, this PR implement the logic of writing row value with checksums, including:
- add `tidb_enable_row_level_checksum` for turn on/off the checksum
- writing checksums on adding/updating record
- writing checksums on ddl backfill columns

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -409,6 +420,23 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx sessionctx.Context,
 				}
 				oldData = append(oldData, value)
 				touched = append(touched, touched[col.DependencyColumnOffset])
+				// the ddl is modify-col
+				if numChecksums > 1 {
```
Comment (cfzjywxk) [inline table/tables/tables.go:424]: Condition like `if extracChecksumNeeded` is easier to understand.

File: table/tables/tables.go [hunk 2]
```diff
@@ -582,6 +633,76 @@ func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, txn kv.Transaction,
...
+	for i, cols := range data {
+		row := rowcodec.RowData{Cols: cols, Data: buf}
+		if !sort.IsSorted(row) {
+			sort.Sort(row)
+		}
+		checksum, err := row.Checksum()
+		buf = row.Data
+		if err != nil {
+			logutil.BgLogger().Warn("skip checksum due to encode error", zap.Int64("tblID", t.meta.ID), zap.Error(err))
+			return nil, buf
```
Comment (cfzjywxk) [inline table/tables/tables.go:689]: Could the returned `buf` be already changed and corrupted?

File: table/tables/tables.go [hunk 3]
```diff
@@ -828,26 +984,54 @@ func (t *TableCommon) AddRecord(sctx sessionctx.Context, r []types.Datum, opts .
 			}
 			row = append(row, value)
 			colIDs = append(colIDs, col.ID)
+			// the ddl is modify-col and the col is write-only or write-reorg
+			if numChecksums > 1 {
```
Comment (cfzjywxk) [inline table/tables/tables.go:988]: Is it possible to abstract the common checksum logic in `AddRecord` and `UpdateRecord` into a utility of `TableCommon`? Not quite sure it looks like there is something similar.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -3332,6 +3332,15 @@ func (d *ddl) AlterTable(ctx context.Context, sctx sessionctx.Context, stmt *ast
...
 		}
 	}
+	if sctx.GetSessionVars().EnableRowLevelChecksum || variable.EnableRowLevelChecksum.Load() {
+		err := dbterror.ErrRunMultiSchemaChanges.GenWithStack("Unsupported multi schema change when row level checksum is enabled")
+		if len(validSpecs) > 1 {
+			return err
+		}
+		if len(validSpecs) == 1 && len(validSpecs[0].NewColumns) > 1 && validSpecs[0].Tp == ast.AlterTableAddColumns {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (zyguan) [reply ddl/ddl_api.go:3342]: https://github.com/pingcap/tidb/pull/36167/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7L3065 There used to be such a function.

File: sessionctx/variable/sysvar.go [hunk 5]
```diff
@@ -1300,6 +1300,15 @@ var defaultSysVars = []*SysVar{
 		}
 		return nil
 	}},
+	{Scope: ScopeGlobal, Name: TiDBEnableRowLevelChecksum, Value: BoolToOnOff(false), Type: TypeBool,
```
[the remark this answers is not in the document]
Comment (zyguan) [reply sessionctx/variable/sysvar.go:1303]: A user session always uses [IsRowLevelChecksumEnabled](https://github.com/pingcap/tidb/pull/43163/files#diff-db5df2c0ea43fa82bdc165fed66584b4c5ba481fb38cf7862889fc8afa64f081R1615) to check whether the checksum is enabled. Changing the global variable won't affect existing sessions.

File: table/tables/tables.go [hunk 6]
```diff
@@ -582,6 +633,76 @@ func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, txn kv.Transaction,
...
+						zap.Int64("tblID", t.meta.ID), zap.Any("cols", t.meta.Columns),
+						zap.Int("origColID", col.ChangeStateInfo.DependencyColumnOffset))
+					return 0, origCol, errors.New("cannot find the original column by change-state-info")
+				}
+			}
+		}
+	}
+	if n := len(nonPubCols); n > 1 {
+		logutil.BgLogger().Warn("skip checksum since the number of non-public columns is greater than 1",
+			zap.Int64("tblID", t.meta.ID), zap.Any("nonPubCols", nonPubCols))
```
[the remark this answers is not in the document]
Comment (zyguan) [reply table/tables/tables.go:668]: Do you mean to print all columns?

File: sessionctx/variable/sysvar.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply sessionctx/variable/sysvar.go:1303]: Would the internal sessions be affected by it?

File: table/tables/tables.go (hunk 2 above)
Comment (zyguan) [reply table/tables/tables.go:689]: The returned `buf` is used to update the cached buf (for reduce memory allocation), its content doesn't matter, all functions that take a `buf` as arg will reset it by `buf = buf[:0]` (if it's not nil).

File: sessionctx/variable/sysvar.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zyguan) [reply sessionctx/variable/sysvar.go:1303]: Yes, I think, `SessionVars.EnableRowLevelChecksum` will be init as global `EnableRowLevelChecksum.Load()` on `NewSessionVars` [here](https://github.com/pingcap/tidb/pull/43163/files#diff-db5df2c0ea43fa82bdc165fed66584b4c5ba481fb38cf7862889fc8afa64f081R1939).

[the remark this answers is not in the document]
Comment (cfzjywxk) [reply sessionctx/variable/sysvar.go:1303]: There was some confusion regarding the management of session variables previously, and we encountered issues where internal sessions were expected or unexpectedly influenced by external variables.

File: table/tables/tables.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply table/tables/tables.go]: Yes, usually the session, transaction, table and column information.

File: table/tables/tables.go (hunk 2 above)
Comment (cfzjywxk) [reply table/tables/tables.go]: How about returning `nil, nil` or adding some comments about it as there are no `mutable or immutable` references in golang, and we've encountered some cases in which the input memory was changed unexpectedly.

Comment (zyguan) [reply table/tables/tables.go]: The returned `buf` has the following properties, thus returning `nil, nil` somehow means invalid this buffer, or we can discard the nil? (do not update the persistent buf when it's nil)

- if cap of buf is enough, return the original buf, maybe with different len
- if cap of buf is not enough, return a new buf with enough capacity (by append(buf, ...)

The content of the returned `buf` is undefined actually, thus callers shouldn't rely on its content (at least for `calcChecksums`). `EncodeRow` also receives a bytes as row buf, and returns the row value contained in the original buf (or a new buf with enough capacity). It's callers' responsiblity to copy the returned bytes if they want to pesist it (membuf has a copy semantics.

**Review by 3AceShowHand — APPROVED** [review]

File: ddl/column.go [hunk 7]
```diff
@@ -1197,12 +1202,33 @@ func newUpdateColumnWorker(sessCtx sessionctx.Context, id int, t table.PhysicalT
 		}
 	}
 	rowDecoder := decoder.NewRowDecoder(t, t.WritableCols(), decodeColMap)
+	checksumNeeded := false
+	failpoint.Inject("forceRowLevelChecksumOnUpdateColumnBackfill", func() {
+		orig := variable.EnableRowLevelChecksum.Load()
+		defer variable.EnableRowLevelChecksum.Store(orig)
```
[the remark this answers is not in the document]
Comment (zyguan) [reply ddl/column.go:1208]: Nope, actually the closure will be rewritten as the following by the failpoint-ctl.
```go
	if _, _err_ := failpoint.Eval(_curpkg_("forceRowLevelChecksumOnUpdateColumnBackfill")); _err_ == nil {
		orig := variable.EnableRowLevelChecksum.Load()
		defer variable.EnableRowLevelChecksum.Store(orig)
		variable.EnableRowLevelChecksum.Store(true)
	}
```

**Review by you06 — APPROVED** [review]

**Pull request merged**